x-pack/filebeat/input/websocket/input.go

```diff
@@ -0,0 +1,438 @@
+	"fmt"
+	"math"
+	"math/rand"
+	"reflect"
+	"strings"
+	"time"
+
+	v2 "github.com/elastic/beats/v7/filebeat/input/v2"
+	inputcursor "github.com/elastic/beats/v7/filebeat/input/v2/input-cursor"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/feature"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-concert/ctxtool"
+	"github.com/elastic/mito/lib"
+
+	"github.com/google/cel-go/cel"
+	"github.com/gorilla/websocket"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+type input struct {
+	time func() time.Time
+	cfg  config
+}
+
+const (
+	inputName string = "websocket"
+	root      string = "state"
+)
+
+func Plugin(log *logp.Logger, store inputcursor.StateStore) v2.Plugin {
+	return v2.Plugin{
+		Name:       inputName,
+		Stability:  feature.Experimental,
+		Deprecated: false,
+		Info:       "Websocket Input",
+		Doc:        "Collect data from websocket api endpoints",
+		Manager:    NewInputManager(log, store),
+	}
+}
+
+func (input) Name() string { return inputName }
+
+func (input) Test(src inputcursor.Source, _ v2.TestContext) error {
+	cfg := src.(*source).cfg
+	if !wantClient(cfg) {
+		return fmt.Errorf("unsupported scheme: %s", cfg.Resource.URL.Scheme)
+	}
+	return nil
+}
+
+// Run starts the input and blocks as long as websocket connections are alive. It will return on
+// context cancellation or type invalidity errors, any other error will be retried.
+func (input) Run(env v2.Context, src inputcursor.Source, crsr inputcursor.Cursor, pub inputcursor.Publisher) error {
+	var cursor map[string]interface{}
+	if !crsr.IsNew() { // Allow the user to bootstrap the program if needed.
+		err := crsr.Unpack(&cursor)
+		if err != nil {
+			return err
+		}
+	}
+	return input{}.run(env, src.(*source), cursor, pub)
+}
+
+func (i input) run(env v2.Context, src *source, cursor map[string]interface{}, pub inputcursor.Publisher) error {
+	cfg := src.cfg
+	i.cfg = cfg
+	log := env.Logger.With("input_url", cfg.Resource.URL)
+
+	metrics := newInputMetrics(env.ID)
+	defer metrics.Close()
+	metrics.resource.Set(cfg.Resource.URL.String())
+	metrics.errorsTotal.Set(0)
+
+	ctx := ctxtool.FromCanceller(env.Cancelation)
+
+	patterns, err := regexpsFromConfig(cfg)
+	if err != nil {
+		metrics.errorsTotal.Inc()
+		return err
+	}
+
+	prg, ast, err := newProgram(ctx, cfg.Program, root, patterns, log)
+	if err != nil {
+		metrics.errorsTotal.Inc()
+		return err
+	}
+	var state map[string]interface{}
+	if cfg.State == nil {
+		state = make(map[string]interface{})
+	} else {
+		state = cfg.State
+	}
+	if cursor != nil {
+		state["cursor"] = cursor
+	}
+
+	// websocket client
+	headers := formHeader(cfg)
+	url := cfg.Resource.URL.String()
+	c, _, err := websocket.DefaultDialer.Dial(url, headers)
+	if err != nil {
+		metrics.errorsTotal.Inc()
+		log.Errorw("failed to establish websocket connection", "error", err)
+		return err
+	}
+	defer c.Close()
+
+	done := make(chan struct{})
+	errChan := make(chan error)
```

Comment:
These could be rolled into a single chan. In all cases of a receive this func returns, the only difference being what it returns. This logic can be placed at the recv based on whether the send was a nil or non-nil error (quietly, no logic is required).
```suggestion
	done := make(chan err)
```

It could conceivably all be rolled into the context with https://pkg.go.dev/context#WithCancelCause, but I don't think that is necessary.